test/blackbox-tests/test-cases/pkg/slang.t

```diff
@@ -0,0 +1,92 @@
+  $ . ./helpers.sh
+
+  $ mkdir dune.lock
+  $ cat >dune.lock/lock.dune <<EOF
+  > (lang package 0.1)
+  > EOF
+
+Helper function to create a lockfile with a given install action and then build it, running the action.
+  $ test_action() {
+  >   dune clean || true
+  >   echo "(install $1)" > dune.lock/test.pkg
+  >   build_pkg test
+  > }
+
+Tests for concat:
+  $ test_action '(run echo (concat foo))'
+  foo
+  $ test_action '(run echo (concat foo bar))'
+  foobar
+  $ test_action '(run echo (concat foo (concat bar (concat baz (concat)))))'
+  foobarbaz
+  $ test_action '(run echo (concat %{pkg-self:name} . %{pkg-self:version}))'
+  test.dev
+  $ test_action '(run (concat e c h o) foo)'
+  foo
+  $ test_action '(run echo foo (concat) bar (concat) baz)' # two spaces between each word because (concat) is the empty string
+  foo  bar  baz
+
+Tests for when:
+  $ test_action '(run echo (when true foo) bar (when false baz) qux)'
+  foo bar qux
+  $ test_action '(run (when (= a b) xxx) (when (and (< 41 42) (<> foo bar)) echo) foo)'
+  foo
+  $ test_action '(run (when invalid-condition echo) foo)'
+  File "dune.lock/test.pkg", line 1, characters 20-37:
+  1 | (install (run (when invalid-condition echo) foo))
+                          ^^^^^^^^^^^^^^^^^
+  Error: This expression is used as a condition and so must evaluate to either
+  "true" or "false", however it evaluated to "invalid-condition".
+  [1]
+  $ test_action '(run (when (and (< 1 2) (concat complex invalid condition)) echo) foo)'
+  File "dune.lock/test.pkg", line 1, characters 33-67:
+  1 | (install (run (when (and (< 1 2) (concat complex invalid condition)) echo) foo))
+                                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+  Error: This expression is used as a condition and so must evaluate to either
+  "true" or "false", however it evaluated to "complexinvalidcondition".
+  [1]
+  $ test_action '(run echo (when (and (< 1 2) (or (concat t r u e) (concat f a l s e))) foo) bar)'
```

Comment:
When is the case of of `(or (concat t r u e) ..)` needed?

Reply:
hopefully never, but boolean literals in blang are just the strings "true" and "false" so this works as a consequence of that.

Reply:
Definitely keep the test case so that we know what's happening. However, I'm a bit concerned by this behavior. Do we really need to implicitly coerce strings to booleans? In other words, do we need this for opam compat.

Reply:
This is already how blangs work so I think we should keep it for consistency with other occurrences of blangs in dune.

Reply:
Yes, that is true, but it is something that I rather dislike and would like to change. We could keep the behavior for now, but I would like to make sure we don't rely on it in any deep way. I want us to introduce booleans (and other values) to `Value.t` and have explicit forms to convert between strings and booleans.

Reply:
Makes sense!